tortoise/backends/asyncpg/client.py

```diff
@@ -95,39 +93,45 @@ def __init__(
         self.extra.pop("fetch_inserted", None)
         self.extra.pop("loop", None)
         self.extra.pop("connection_class", None)
+        self.pool_minsize = int(self.extra.pop("minsize", 1))
+        self.pool_maxsize = int(self.extra.pop("maxsize", 5))
 
         self._template: dict = {}
-        self._connection: Optional[asyncpg.Connection] = None
-        self._lock = asyncio.Lock()
+        self._pool: Optional[asyncpg.pool] = None
+        self._lock = asyncio.Lock()  # only used for isolating units of work in transactions
```

Comment:
You don't want this shared, as it essentially serialises all operations. We need a separate Lock instance per connection.
They are cheap to make, so I'd create a new instance in the TransactionWrapper constructor.

Look at this commit: https://github.com/tortoise/tortoise-orm/pull/234/commits/a562a2a32789b70bccadff29ac1a66b1cdc95d39